packages/react-menu/src/components/Menu/useMenu.tsx

```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { makeMergeProps, resolveShorthandProps, useMergedRefs, useControllableValue } from '@fluentui/react-utilities';
+import { MenuProps, MenuState } from './Menu.types';
+import { MenuTrigger } from '../MenuTrigger/index';
+import { MenuList } from '../MenuList/index';
+
+export const menuShorthandProps: (keyof MenuProps)[] = ['menuPopup'];
+
+const mergeProps = makeMergeProps<MenuState>({ deepMerge: menuShorthandProps });
+
+/**
+ * Create the state required to render Menu.
+ *
+ * The returned state can be modified with hooks such as useMenuStyles,
+ * before being passed to renderMenu.
+ *
+ * @param props - props from this instance of Menu
+ * @param ref - reference to root HTMLElement of Menu
+ * @param defaultProps - (optional) default prop values provided by the implementing type
+ *
+ * {@docCategory Menu }
+ */
+export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultProps?: MenuProps): MenuState => {
+  const state = mergeProps(
+    {
+      ref: useMergedRefs(ref, React.useRef(null)),
+      menuPopup: { as: 'div' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, menuShorthandProps),
+  );
+
+  // TODO Better way to narrow types ?
+  const children = React.Children.toArray(state.children) as React.ReactElement[];
+
+  // TODO throw warnings in development safely
+  if (children.length !== 2) {
+    // eslint-disable-next-line no-console
+    console.warn('Menu can only take one MenuTrigger and one MenuList as children');
+  }
+
+  children.forEach(child => {
+    if (child.type === MenuTrigger) {
+      state.menuTrigger = child;
+    }
+
+    if (child.type === MenuList) {
```

Comment:
```suggestion
    if (child.type !== MenuTrigger) {
```

We cannot rely that customer will always pass us `MenuList`, we should take as a contract that user will pass us `MenuTrigger` (utility component) and _some other component with menu_. Otherwise, we will block this scenario:

```tsx
<Menu>
  <MenuTrigger /> // this is okay, as it renders nothing and cannot be actually customized
  <MyCustomMenuList /> // we cannot require MenuList to be always a child as it blocks customization
</Menu>
```

Reply:
yeah that's a good catch